// clubhouse-lib/build/TokenRequestFactory.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var TokenRequestFactory = function () {
  function TokenRequestFactory(token) {
    _classCallCheck(this, TokenRequestFactory);

    this.token = token;
  }

  _createClass(TokenRequestFactory, [{
    key: 'createRequest',
    value: function createRequest(url) {
      var method = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'GET';
      var body = arguments[2];

      var urlWithToken = url + '?token=' + this.token;
      var headers = {
        Accept: 'application/json',
        'Content-Type': 'application/json; charset=utf-8'
      };

      return new Request(urlWithToken, {
        body: JSON.stringify(body),
        headers: headers,
        method: method
      });
    }
  }]);

  return TokenRequestFactory;
}();

exports.default = TokenRequestFactory;
